ykrt/src/mt.rs

```diff
@@ -294,8 +294,9 @@ impl MTThread {
         // to update values, allowing us to detect if we were interrupted. If we were interrupted,
         // we simply retry the whole operation.
 
-        // We need Acquire ordering, as PHASE_COMPILED will need to read information written to
-        // external data as a result of the PHASE_TRACING -> PHASE_COMPILED transition.
+        // We need Acquire ordering, as PHASE_COMPILING and PHASE_COMPILED need to read information
+        // written to external data. Alternatively, this load could be Relaxed but we would then
+        // need to place an Acquire fence in PHASE_COMPILING and PHASE_COMPILED.
```

Comment:
I realised that I'd never looked in to how memory fences work.

I'm not sure about this, but it looks to me like they'd perform worse, because they cause *all* memory operations to synchronise, whereas an atomic value only synchronises WRT one specific address?

Looking at the intel manual:

![image](https://user-images.githubusercontent.com/604955/102073900-e6036580-3dfb-11eb-8f64-84b4dc94880e.png)


Here "weakly ordered memory types" are stuff like acquire and release, I think?

Reply:
Wait, what do you mean by "Acquire fence" actually?

Reply:
`SeqCst` atomic operations synchronize with all other memory operations I think. `Acquire` and `Release` atomic operations synchronize with some other memory operations. `Relaxed` atomic operations only synchronize with other atomic operations on the same memory location.

As far as I know fences upgrade other atomic operations to a less weak memory ordering.

According to https://doc.rust-lang.org/stable/std/sync/atomic/fn.fence.html:

> Depending on the specified order, a fence prevents the compiler and CPU from reordering certain types of memory operations around it. That creates synchronizes-with relationships between it and atomic operations or fences in other threads.
>
> A fence 'A' which has (at least) Release ordering semantics, synchronizes with a fence 'B' with (at least) Acquire semantics, if and only if there exist operations X and Y, both operating on some atomic object 'M' such that A is sequenced before X, Y is synchronized before B and Y observes the change to M. This provides a happens-before dependence between A and B.

Reply:
> As far as I know fences upgrade other atomic operations to a less weak memory ordering.

Right, so it synchronises *more*, which may be slower, right?

But I'm wondering if @ltratt meant an actual memory fence, or he just used the term "fence" casually when talking about a regular atomic op.

Reply:
In this case the load would be relaxed, but when the state turned out to be `PHASE_COMPILING` or `PHASE_COMPILED` there would be a fence that as far as I know upgrades the previous store and **only** the previous store to an acquire load.

Reply:
> In this case the load would be relaxed, but when the state turned out to be PHASE_COMPILING or PHASE_COMPILED there would be a fence that as far as I know upgrades the previous store and only the previous store to an acquire load.

Exactly. [I expect this would be slower, overall, but until we benchmark it we can't know for sure.]

Reply:
Agreed.